tls/_common/_constructs.py

```diff
@@ -211,6 +213,35 @@ def EnumSwitch(type_field, type_enum, value_field, value_choices,  # noqa
                              default=default))
 
 
+class TLSExprValidator(construct.Validator):
+    """
+    Like :py:class:`construct.ExprValidator`, but raises a
+    :py:class:`tls.exceptions.TLSValidationException` on validation failure.
+    """
+    def __init__(self, subcon, validator):
+        super(TLSExprValidator, self).__init__(subcon)
+        self._validate = validator
+
+    def _decode(self, obj, context):
+        if not self._validate(obj, context):
+            raise TLSValidationException("object failed validation", obj)
+        return obj
+
+
+def TLSOneOf(subcon, valids):  # noqa
+    """
+    Validates that the object is one of the listed values, both during parsing
+    and building. Like :py:meth:`construct.OneOf`, but raises a
+    :py:class:`tls.exceptions.TLSValidationException` instead of a
+    ``ConstructError`` subclass on mismatch.
+
+    This is necessary because any ConstructError signifies the end of
```

Comment:
It might be nice to put an explanation like this into `TLSExprValidator`'s doc string.